React UI: Add Starting Screen

LeviHarrison opened the pull request:
<!--
    Don't forget!
    
    - If the PR adds or changes a behaviour or fixes a bug of an exported API it would need a unit/e2e test.
    
    - Where possible use only exported APIs for tests to simplify the review and make it as close as possible to an actual library usage.
    
    - No tests are needed for internal implementation changes.
    
    - Performance improvements would need a benchmark test to prove it.
    
    - All exposed objects should have a comment.
    
    - All comments should start with a capital letter and end with a full stop.
 -->

This PR adds the `api/v1/status/walreplay` endpoint that tracks the WAL replay in real-time, and a page in the UI that is rendered when the server is starting up.

![with-logo](https://user-images.githubusercontent.com/54278938/112852291-caa22a80-9079-11eb-9a65-6f548c5e8c21.gif)

And without the logo:

![without-logo](https://user-images.githubusercontent.com/54278938/112854224-a0516c80-907b-11eb-9f2d-0d2f1aca779b.gif)

juliusv commented on web/ui/react-app/src/pages/starting/Starting.tsx:
```diff
@@ -0,0 +1,59 @@
+
+  return (
+    <div className="text-center m-3">
+      <Logo height="400" />
```
Hmm, it's a nice touch, but I'm not sure we should include a logo here, as we don't do that anywhere else. Of course, that's something we can think about separately some time on other pages, but let's not introduce it now just for a page that most people will never get to see.

juliusv commented on web/ui/react-app/src/pages/starting/Starting.tsx:
```diff
@@ -0,0 +1,59 @@
+  return (
+    <div className="text-center m-3">
+      <Logo height="400" />
+      <div className="m-4">
```
Could you share a screenshot in the PR of what this page looks like when the WAL is loading?

LeviHarrison replied:
Just updated the comment with a (sadly low quality) GIF.

juliusv commented on web/ui/react-app/src/pages/starting/Starting.tsx:
```diff
@@ -0,0 +1,59 @@
+        {status?.started && !status?.done ? (
+          <div className="mx-5">
+            <p>
+              Replaying Wal ({status?.read - status?.first}/{status?.last - status?.first})
```
```suggestion
              Replaying WAL ({status?.read - status?.first}/{status?.last - status?.first})
```

juliusv commented on web/api/v1/api.go:
```diff
@@ -1351,6 +1352,25 @@ func (api *API) serveTSDBStatus(*http.Request) apiFuncResult {
+	Done    bool `json:"done"`
+}
+
+func (api *API) serveWalReplayStatus(w http.ResponseWriter, r *http.Request) {
```
```suggestion
func (api *API) serveWALReplayStatus(w http.ResponseWriter, r *http.Request) {
```

juliusv commented on tsdb/head.go:
```diff
@@ -53,13 +53,23 @@ var (
 	// rolled back or committed.
 	ErrAppenderClosed = errors.New("appender closed")
+
+	WalReplayStatus walReplayStatus
```
See https://github.com/golang/go/wiki/CodeReviewComments#initialisms for capitalization of initialisms like WAL.

juliusv reviewed (commented):
Thanks, looking good in many ways already, but some comments :)

LeviHarrison commented on docs/querying/api.md:
```diff
@@ -1031,6 +1031,34 @@ $ curl http://localhost:9090/api/v1/status/tsdb
 
 *New in v2.15*
 
+### WAL Replay Stats
```
Yeah, I was wondering what release this would fall into. I think it's probably `v2.26`, although I'm not exactly sure how your release cycles work and there's already an rc out.

juliusv replied:
Since it's a new feature and not a bug fix, it'd go only into 2.27, since the 2.26 RC is already cut.

LeviHarrison commented on docs/querying/api.md:
````diff
@@ -1031,6 +1031,34 @@ $ curl http://localhost:9090/api/v1/status/tsdb
+}
+```
+
+NOTE: This endpoint is available before the server has been marked ready and is updated in real-time to facilitate monitoring the progress of the WAL replay.
````
Yup, that seems to be right. Grammarly probably likes its hyphens a little bit too much 😄

LeviHarrison commented on web/ui/react-app/src/hooks/useFetch.ts:
```diff
@@ -32,3 +57,67 @@ export const useFetch = <T extends {}>(url: string, options?: RequestInit): Fetc
+        // The server sends back a 503 if it isn't ready,
+        // if we get back anything else that means something has gone wrong.
+        if (res.status !== 503) {
+          setIsResponding(false);
```
My concern was with proxies, which still respond even if the endpoint isn't available, and those responses aren't handled as exceptions. I also now realize that `isResponding` (or something like it) should be set to false in the exception handler too, for instances that aren't behind proxies or other general network issues. Maybe just a blanket `isUnexpected` which can be used for non `200` and `503` status codes as well as network errors.

juliusv replied:
Yep, something like that sounds fine I guess.

LeviHarrison commented on web/ui/react-app/src/pages/starting/Starting.tsx:
```diff
@@ -25,10 +25,10 @@ export const StartingContent: FC<StartingContentProps> = ({ status, isResponding
       <div className="m-4">
         <h2>Starting up...</h2>
-        {status?.started && !status?.done ? (
+        {status?.started ? (
```
I just changed this so the loading bar won't un render once the replaying is done, and instead will turn green.

LeviHarrison commented on tsdb/head.go:
```diff
@@ -53,13 +53,23 @@ var (
 	// rolled back or committed.
 	ErrAppenderClosed = errors.New("appender closed")
+
+	WalReplayStatus walReplayStatus
```
The method you're referring to relies on this data structure:

https://github.com/prometheus/prometheus/blob/bbc096cace886a5ca53cde72f56a36181e98d0b7/cmd/prometheus/main.go#L1056-L1062

Which is only populated with the `Head` once the instance is ready, and therefore only allows us to serve WAL replay status after it's already over. 

The current approach I'm leaning towards is having a variable in `head.go` like before, but having `api.go` call a function in `head.go` that reads it (after obtaining the lock), instead of having `api.go` read it directly like it currently does. This is the same way the variable containing the `readyStorage` value is handled.

Thoughts?

juliusv replied:
Ah, that makes things more complicated. Definitely we should solve this without globals though.

I wonder if we could restructure the whole readiness handling such that `readyStorage` holds the TSDB instance from the beginning (so `db` would never be `nil`), but instead its wrapper methods would determine readiness by calling the `WALReplayStatus()` method on the TSDB. Then I'd also extend the `TSDBAdminStats` interface at https://github.com/prometheus/prometheus/blob/9549a15c6ffd1e997ee15f6262e5a215a64d5661/web/api/v1/api.go#L150-L157 with a new `WALReplayStatus()` method and implement that in `readyStorage` as well. Just that that method would not have the usual readiness handling that `Stats()` etc. have, since it's already supposed to work before the TSDB is ready. Then you could use that method from the web API.

What do you think?

roidelapluie replied:
I think that this would be too much. I think we handle this differently without going to the global or changing the interfaces. This is very prometheus-tsdb specific.

LeviHarrison replied:
Oh sorry @roidelapluie I hadn't refreshed my page in a while before I submitted that comment.

I don't think that it would be too difficult, only requiring changes to a few functions/structers (`tsdb.Open`, `tsdb.open`, `*readyStorage.get`, `main.openDBWithMetrics`, `api.TSDBAdminStats` is all I can think of) and it shouldn't affect anything else beyond   `main.go`.

That being said, I would definitely be open to exploring other approaches.

juliusv replied:
Thanks for the exploration & explanation! Yeah, generally, splitting the TSDB creation and initialization into two externally visible steps could be a decent way to go. Another option would be to pass a stats object into the `tsdb.Open()` call that is updated by the DB initialization, but can be read simultaneously by the caller.

Since we definitely need to touch the TSDB code in one way or another, I would just like to make sure that the TSDB maintainers @codesome and @bwplotka are aware of this. Maybe they also have a specific opinion / advice on how to most cleanly expose initialization progress.

bwplotka replied:
Ack, looking on this part.

bwplotka replied:
I checked and ideally we get rid of the global.

@juliusv idea sounds quite accessible.

>  Another option would be to pass a stats object into the tsdb.Open() call that is updated by the DB initialization, but can be read simultaneously by the caller.

Otherwise, I would pass an additional function/interface to `API` that will have a direct connection to WAL/Head. Head can have `WalReplayStatus()` and we directly pass this method during initialization in Prometheus cmd/prometheus/main.go - no need to go through readiness storage 🤗

roidelapluie commented:
Thank you for this! This looks awesome.

I have a bug before the first segment is replayed:
![-42](https://user-images.githubusercontent.com/291750/112883445-973dbb00-90ce-11eb-8540-9d0c82e0de8d.png)

I would like the progress bar to take less than the full width of the screen. I think this is too wide.

roidelapluie commented:
If Prometheus is stopped while replaying the wall and comes back, the progressbar is not reset to 0 ( but it might be because of the bug in the previous comment )
![wall](https://user-images.githubusercontent.com/291750/112884182-7c1f7b00-90cf-11eb-9cb2-196ca503693c.png)

LeviHarrison commented:
I'm fairly certain I found the bug with the progress bar values, but I'll have to wait to push until I finish up with the WAL replay stuff on the server.

In the meantime, I changed the width:

![Screenshot from 2021-03-29 16-02-31](https://user-images.githubusercontent.com/54278938/112893691-030d2d00-90a9-11eb-9fd9-9084f740faed.png)

How's this?

roidelapluie commented:
I am wondering is a bar of the same size as the text would make sense.

LeviHarrison commented:
Ok, here it is:

![Screenshot from 2021-03-29 16-37-44](https://user-images.githubusercontent.com/54278938/112897076-2c2fbc80-90ad-11eb-954e-61bb529bfed1.png)

LeviHarrison commented:
Also, on a whim, I noticed that when I `^C` (`SIGINT`) Prometheus in my terminal during the WAL replay, it will stop everything else but keep the replay going (ex: API calls stop responding). This is expected behavior, right?

roidelapluie commented:
> Also, on a whim, I noticed that when I `^C` (`SIGINT`) Prometheus in my terminal during the WAL replay, it will stop everything else but keep the replay going (ex: API calls stop responding). This is expected behavior, right?

Mmmh I am not an expert here but at least your PR is not changing this behaviour from what I see.

LeviHarrison commented:
Nope, I just noticed it because it prevents me from restarting the instance right after stopping it while testing.

roidelapluie commented:
for the size, can we apply the css classes `col-2 mx-auto px-0` ?

LeviHarrison commented:
Sure can!

![Screenshot from 2021-03-29 17-40-57](https://user-images.githubusercontent.com/54278938/112903839-0fe44d80-90b6-11eb-9626-506526e5a136.png)

roidelapluie commented:
cc @codesome @bwplotka can you look at the tsdb questions in this issue?

bwplotka commented on tsdb/head.go:
```diff
@@ -53,13 +53,23 @@ var (
 	// rolled back or committed.
 	ErrAppenderClosed = errors.New("appender closed")
+
+	WALReplayStatus walReplayStatus
```
Is this global variable? I think that's might be not the best option. First of all, it will block people for running multiple Heads (e.g for test or any other purposes).

In fact we do this all the time in Thanos and Cortex projects.

LeviHarrison replied:
Yeah, in my working version of the codebase I had moved this into a field in the `Head` struct, added functions to modify the values that were concurrency safe, and also added the `WalReplayStats` (indirectly) to the `API` struct. I was just waiting on getting guidance on the head configuration thing so I could complete the work. Probably should have pushed as is - sorry for the trouble.

bwplotka replied:
ah, perfect, looing now!

bwplotka commented on tsdb/head.go:
```diff
@@ -750,6 +760,10 @@ func (h *Head) Init(minValidTime int64) error {
 		return errors.Wrap(err, "finding WAL segments")
 	}
 
+	WALReplayStatus.Started = true
```
I think there is a race here. What if API is reading from the variable in the moment it's writing? Unfortunately we might want to add some concurrency control (mutexes)

bwplotka commented on web/api/v1/api.go:
```diff
@@ -1351,6 +1352,25 @@ func (api *API) serveTSDBStatus(*http.Request) apiFuncResult {
 	}, nil, nil, nil}
 }
 
+type walReplayStatus struct {
```
This struct is already defined once in head, can we reuse it? 🤗

LeviHarrison replied:
My issue here was the JSON mappings. They're required for the endpoint to work and I didn't know if we wanted them in structs not defined specifically for the API. Also, I wonder how the mutex would decode?. It might end up being private anyway, but I'm not sure.

LeviHarrison replied:
Also, now we have the `WALReplayState`, which should probably be returned by the API in its string format, which requires a different struct definition.

bwplotka replied:
We can add mapping TBH to anything, but I think it's common for our trust to reproduce those, so not a big deal

bwplotka commented on web/api/v1/api.go:
```diff
@@ -1351,6 +1352,25 @@ func (api *API) serveTSDBStatus(*http.Request) apiFuncResult {
+	httputil.SetCORS(w, api.CORSOrigin, r)
+	api.respond(w, walReplayStatus{
+		First:   tsdb.WALReplayStatus.First,
+		Last:    tsdb.WALReplayStatus.Last,
```
As mentioned before, I would suggest actually passing this through API struct and add mutex (sync.Lock) to avoid undefined behaviours and even crashes.

bwplotka reviewed (changes requested):
Thank you for this contribution! 💪🏽  I reviewed mostly the TSDB code and unfortunately, we need to add locking and move the status variable out of globals to make the API reliable. Let me know if you agree or I missed something.

Otherwise, it looks quite epic and will be super useful for everyone 🤗

LeviHarrison commented on docs/querying/api.md:
```diff
@@ -1031,6 +1031,36 @@ $ curl http://localhost:9090/api/v1/status/tsdb
+
+**first**: The first segment needed to be replayed.
+**last**: The last segment needed to be replayed.
+**read**: The latest segment replayed.
```
`read` is the most recent segment to be replayed. So if we start at the 5th segment, then read it, the `read` value would be 5. In the front-end code, the `first` value is subtracted from the `read` value to generate the progress bar. I now realize this is flawed because the `read` value can be the same as the `first` value, showing no progress, even though that just means that only the first segment _has_ been read. The `read` value not being set to the starting value is what caused the bugs that @roidelapluie found. I think I'll probably need to rework this, maybe just making it the total number of segments read regardless of the ID, starting from 0, and then let the client (or server) figure the progress out.

LeviHarrison replied:
Furthermore, the reason that all of these fields are so specific (as to a single number counting from 0) is that I wanted this endpoint to provide the same data as the log lines do.

bwplotka replied:
Yea, I think we don't need to be strict towards log line, will look how it looks now

LeviHarrison commented on docs/querying/api.md:
```diff
@@ -1031,6 +1031,36 @@ $ curl http://localhost:9090/api/v1/status/tsdb
+{
+  "status": "success",
+  "data": {
+    "first": 1076,
```
I like it! Should that be an addition, or completely replace the `first`, `read`, and last fields? I intentionally wrote this endpoint to expose more information than just the starting page needs in case anyone else would like to access the data, but I could support either direction.

bwplotka replied:
Replace IMO (:

LeviHarrison commented:
Thanks so much for your input @bwplotka :) I just left a few responses, and I'll get to working as soon as I can!

LeviHarrison commented on docs/querying/api.md:
````diff
@@ -1031,6 +1031,36 @@ $ curl http://localhost:9090/api/v1/status/tsdb
+GET /api/v1/status/walreplay
+```
+
+**first**: The first segment needed to be replayed.
````
Since the WAL replay doesn't always start at the first segment, `first` is the position (starting from 1) of the first segment that needs to be read.

LeviHarrison commented:
Ok, that should be all of it! I added the `DBStats` struct, which is passed into `Open` and furthermore into `NewHead` where it can be accessed during the WAL replay. I also changed the API response to be more succinct.

LeviHarrison commented:
I'm pretty sure the current `test_tsdb_go115` failure is a flake.

roidelapluie commented:
Could you rebase? Thanks

LeviHarrison commented:
Gotcha!

LeviHarrison commented:
I plan to add tests once the implementation of the head stats object has been agreed upon, so I can avoid changing them around a bunch after reviews.

LeviHarrison commented:
It seems that this is going to miss 2.27.

bwplotka commented on cmd/prometheus/main.go:
```diff
@@ -1083,14 +1090,29 @@ func (s *readyStorage) Set(db *tsdb.DB, startTimeMargin int64) {
-// get is internal, you should use readyStorage as the front implementation layer.
+// SetStats sets the stats object,
+// which should be available before the DB is finished opening.
+func (s *readyStorage) SetStats(stats *tsdb.DBStats) {
```
Do we need this method? Why not passing this through constructor always?

bwplotka replied:
or even better, created through constructor?

LeviHarrison replied:
If you mean adding it to the variable definition over here instead of using that function, I'm pretty sure that will work.

https://github.com/prometheus/prometheus/blob/9a19336b29060eb0414e80382b904777a002e91c/cmd/prometheus/main.go#L436-L437

bwplotka replied:
Will work? Can we do this? Limiting number of methods is always helpful.

LeviHarrison replied:
Yup!

bwplotka commented on docs/querying/api.md:
```diff
@@ -1033,6 +1033,39 @@ $ curl http://localhost:9090/api/v1/status/tsdb
+  "status": "success",
+  "data": {
+    "read": 2,
+    "total": 5
```
```suggestion
    "total": 5,
```

bwplotka commented on docs/querying/api.md:
```diff
@@ -1033,6 +1033,39 @@ $ curl http://localhost:9090/api/v1/status/tsdb
+  "data": {
+    "read": 2,
+    "total": 5
+    "progress": 40,
```
Do we need `read` if we have this progress?

bwplotka commented on docs/querying/api.md:
```diff
@@ -1033,6 +1033,39 @@ $ curl http://localhost:9090/api/v1/status/tsdb
+    "read": 2,
+    "total": 5
+    "progress": 40,
+    "state": "in progress"
```
To ensure best API it's best to ensure the API is consistent and minimal. I still think it exposes too much information. As is there are multiple ways of getting the same information.

E.g to learn that the WAL replay is in progress I could:
* Check the percentage is higher than 0 and lower than 100
* I could check that `read` is higher than 0 and lower than `total`
* I could check the replay `state` == `in progress`

Having multiple ways of doing same thing for UI when talking to humans, sometimes makes sense, but for programmatic API it should be avoided. Every field adds extra cost. Adds and creates complexity, migration issues, documentation burden, support burden etc. 

Do you think we could aim for one minimal set of field that tells us everything?

I would really focus on literally one, two maybe 3 fields e.g:

```
status: "success", 
data:  {
  start_time: "<timestamp of replay start>"
  progress_percentage: 40,
  message: "replaying segment file 1", or doing X, doing Y, waiting for replay to start, parsing, finished - some human readable message

```

roidelapluie replied:
message: "replaying segment file 1", or doing X, doing Y, waiting for replay to start, parsing, finished - some human readable message


Not really since we want to show a progressbar.

    "min": 210,
    "max": 271
    "current": 250,
    
   might be better

LeviHarrison replied:
Although `min`/`max`/`current` is a lot of fields, I think it's the best format because it provides machine-readable information about the magnitude of the replay not just for the starting screen but for other systems that might choose to process that data instead of just displaying it to a user (making the message less feasible). At the same time, I think we'll probably have to add in that state enum because checking those values alone isn't a reliable way to determine the status of the replay (if there are not segments, all three could be 0). I like the idea of `start_time` too, but maybe that's too many fields...

bwplotka replied:
So message is nice as it might tell what replay is focusing now, just nice to have. But yes not necessarily

LeviHarrison replied:
So should we add that on too, or just leave it at this?

bwplotka replied:
I would be against concrete number because:
* We may do things concurrently so suddenly number of segments will totally not aligned.
* We may do extra additional stages before, after or in between. 

I feel simple percentage allows us to decide on server side whatever percentage means and is the only thing that user needs as it's just loading bar, no?

roidelapluie replied:
This API is not covered by our stability guarantees. I like the fact that the UI shows what we have in the logs.

LeviHarrison replied:
The think the one thing a percentage is missing is the magnitude of segments that need to be loaded. 40% could be 4/10 or 400/1000; the ladder will take significantly longer. Maybe we just include the total number of segments that need to be loaded, and leave out the min and current. I think it will be nice for a user to be able to estimate how much longer it will take. On the other hand, I do see your point with the alignment and extra stages.

bwplotka replied:
Cool, I think we find a good middleground (:

bwplotka commented on tsdb/db.go:
```diff
@@ -574,13 +589,16 @@ func validateOpts(opts *Options, rngs []int64) (*Options, []int64) {
 }
 
-func open(dir string, l log.Logger, r prometheus.Registerer, opts *Options, rngs []int64) (_ *DB, returnedErr error) {
+func open(dir string, l log.Logger, r prometheus.Registerer, opts *Options, rngs []int64, stats *DBStats) (_ *DB, returnedErr error) {
```
Why not just returning one?

LeviHarrison replied:
The whole problem in the first place is that the `open` function triggers the WAL replay, and we have to wait for that to finish until we get the return. So if we return the stats object, that means we can't get any of the WAL replay progress until it's actually completed, which defeats the whole point.

bwplotka replied:
Ah you are right, thanks. Alternatively we could return unready DB, but not worth it (:

bwplotka commented on tsdb/head.go:
```diff
@@ -306,6 +308,64 @@ func newHeadMetrics(h *Head, r prometheus.Registerer) *headMetrics {
+}
+
+// GetWALReplayStatus returns the WAL replay status information.
+func (s *WALReplayStatus) GetWALReplayStatus() *WALReplayStatus {
```
Let's make sure we copy struct (so just return non pointer). We don't want to put pointer as it's still the same pointer to the same WALReplayStatus object so locking does not help ): 
```suggestion
func (s *WALReplayStatus) GetWALReplayStatus() WALReplayStatus {
```

LeviHarrison replied:
Govet complains about copying the lock value here (see failed tests)

bwplotka commented on tsdb/head.go:
```diff
@@ -327,7 +387,7 @@ func (h *Head) PostingsCardinalityStats(statsByLabelName string) *index.Postings
 
 // NewHead opens the head block in dir.
-func NewHead(r prometheus.Registerer, l log.Logger, wal *wal.WAL, opts *HeadOptions) (*Head, error) {
+func NewHead(r prometheus.Registerer, l log.Logger, wal *wal.WAL, opts *HeadOptions, stats *HeadStats) (*Head, error) {
```
Can head just give us the stats with some `GetStats()`? Why we need to compose it and create from outside? It's just same instance always and we don't need to be preffiled (we can make head being the owner of initialising this object)

LeviHarrison replied:
We don't get the DB object with the head back until after `Open` has returned, and thereby the WAL replay has been completed.

bwplotka reviewed (commented):
Looks good, just 2 important things one around HTTP API second around programmatic API.

Thanks!

LeviHarrison commented:
Sorry to anyone whose been pinged for code review, I accidentally messed up a rebase

LeviHarrison commented:
I've changed the API response to `min`/`max`/`current`, but I suspect this is not over yet :) Again, I'll bring up _not_ passing the `DBStats` object to the main `tsdb.Open` method, and instead creating a separate `(*DBStats) Open(...)`. I've seen that `tsdb.Open` is widely used outside of Prometheus, and I just don't think it's worth it forcing everyone to add an extra `nil` for something they most likely won't use.

bwplotka commented:
Thanks a lot! Looking now

bwplotka commented on cmd/prometheus/main.go:
```diff
@@ -1193,6 +1203,14 @@ func (s *readyStorage) Stats(statsByLabelName string) (*tsdb.Stats, error) {
 	return nil, tsdb.ErrNotReady
 }
 
+// WALReplayStatus implements the api_v1.TSDBStats interface.
```
```suggestion
// WALReplayStatus implements the github.com/prometheus/prometheus/web/api/v1.TSDBStats interface.
```

LeviHarrison replied:
The other function implementations don't though

bwplotka replied:
Nothing is perfect, feel free to ignore this then or fix the other too (:

bwplotka commented on tsdb/head.go:
```diff
@@ -307,6 +309,35 @@ func newHeadMetrics(h *Head, r prometheus.Registerer) *headMetrics {
+// GetWALReplayStatus returns the WAL replay status information.
+func (s *WALReplayStatus) GetWALReplayStatus() WALReplayStatus {
+	s.RLock()
+	x := s
```
```suggestion
	x := *s
```

Otherwise we still might have race when dereferencing on line 338 🙈 

But it's much better than it was!

bwplotka replied:
Actually linter already mention this: 
![image](https://user-images.githubusercontent.com/6950331/120528521-772ebe80-c3d3-11eb-9fe9-0ba9e2665948.png)

Robots take my job away 😢

LeviHarrison replied:
It (still) complains about copying the lock value, is that a problem?

LeviHarrison replied:
Yeah, the linter is complaining about it now. Those darn robots...

bwplotka reviewed (approved):
I think this is ready to be merged, except for super minor nits! Thanks for this and for the patience! 💪🏽

LeviHarrison commented:
Thanks so much! Before we merge, I think I should also probably add tests for the `DBStats` stuff, as well as the API route. And I'm still a little bit uneasy with the impact this may have on downstream projects:

> I'll bring up not passing the DBStats object to the main tsdb.Open method, and instead creating a separate (*DBStats) Open(...). I've seen that tsdb.Open is widely used outside of Prometheus, and I just don't think it's worth it forcing everyone to add an extra nil for something they most likely won't use.

bwplotka reviewed (approved):
I think extra param is not ideal, but we can live with that.

This PR LGTM (: Thanks for hard work here 💪🏽 

cc @juliusv @roidelapluie do you want to take another look?

bwplotka commented on tsdb/head.go:
```diff
@@ -307,6 +309,35 @@ func newHeadMetrics(h *Head, r prometheus.Registerer) *headMetrics {
+	x := *s
+	s.RUnlock()
+
+	return x
```
This is still failing although there is nothing wrong given the WALReplayStatus does not  have pointers inside itself (that would not work):

```
GO111MODULE=on /go/bin/golangci-lint run --timeout 2m ./...
tsdb/head.go:335:7: copylocks: assignment copies lock value to x: github.com/prometheus/prometheus/tsdb.WALReplayStatus (govet)
	x := *s
	     ^
tsdb/head.go:338:9: copylocks: return copies lock value: github.com/prometheus/prometheus/tsdb.WALReplayStatus (govet)
	return x
```

I would suggest explicit copy then. Create new `WALReplayStatus` and copy fields manually. WDYT?

LeviHarrison replied:
Sure!

bwplotka commented:
Let's merge and iterate if something is missing, it looks good for me! (:

juliusv commented:
Thanks a bunch again!

Just testing this again (added a `time.Sleep()` in the Go code to simulate a slowly-starting storage). One thing I noticed: the starting screen is shown on initial page load, but since the `App` itself is only rendered once upon the initial load, it does not show the starting screen when I then further navigate to any other pages, like /targets:

![notready-error](https://user-images.githubusercontent.com/538008/120896406-a8043300-c621-11eb-9df1-a09230fcb047.png)

It's ok, but it would be even nicer if we could either prevent subsequently navigating to any specific pages (that don't work yet) or show a starting indicator on each of them, rather than an error. Further, the current behavior is that any page redirects to `/starting`, and when ready, redirects to `/`, instead of to the original page someone navigated to. A future improvement could indeed be to just have the starting spinner on all relevant pages and then stay on that page once the server is ready.

LeviHarrison commented:
Makes sense, I'll add that when I can.

roidelapluie commented:
Awesome!!

LeviHarrison commented:
@juliusv #8909 fixes that.
